File: lib/models/election.dart
class Election {
  final String id;
  final String title;
  final String description;
  final String organizerEmail;
  final String? organizationId;
  final DateTime startDate;
  final DateTime endDate;
  final String status;
  final String resultVisibility;
  final List<Candidate> candidates;
  final List<Voter> voters;
  final int totalVotes;
  final bool? hasVoted;
  final DateTime createdAt;
  final DateTime? closedAt;

  Election({
    required this.id,
    required this.title,
    required this.description,
    required this.organizerEmail,
    this.organizationId,
    required this.startDate,
    required this.endDate,
    required this.status,
    required this.resultVisibility,
    required this.candidates,
    required this.voters,
    required this.totalVotes,
    this.hasVoted,
    required this.createdAt,
    this.closedAt,
  });

  factory Election.fromJson(Map<String, dynamic> json) {
    return Election(
      id: json['_id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      organizerEmail: json['organizer_email'] ?? '',
      organizationId: json['organization_id'],
      startDate: DateTime.parse(json['start_date'] ?? DateTime.now().toIso8601String()).toLocal(),
      endDate: DateTime.parse(json['end_date'] ?? DateTime.now().toIso8601String()).toLocal(),
      status: json['status'] ?? 'draft',
      resultVisibility: json['result_visibility'] ?? 'hidden',
      candidates: (json['candidates'] as List<dynamic>? ?? [])
          .map((c) => Candidate.fromJson(c))
          .toList(),
      voters: (json['voters'] as List<dynamic>? ?? [])
          .map((v) => Voter.fromJson(v))
          .toList(),
      totalVotes: json['total_votes'] ?? 0,
      hasVoted: json['has_voted'],
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()).toLocal(),
      closedAt: json['closed_at'] != null ? DateTime.parse(json['closed_at']).toLocal() : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'description': description,
      'organizer_email': organizerEmail,
      'organization_id': organizationId,
      'start_date': startDate.toIso8601String(),
      'end_date': endDate.toIso8601String(),
      'status': status,
      'result_visibility': resultVisibility,
      'candidates': candidates.map((c) => c.toJson()).toList(),
      'voters': voters.map((v) => v.toJson()).toList(),
      'total_votes': totalVotes,
      'has_voted': hasVoted,
      'created_at': createdAt.toIso8601String(),
      'closed_at': closedAt?.toIso8601String(),
    };
  }

  bool get isDraft => status == 'draft';
  bool get isActive => status == 'active';
  bool get isClosed => status == 'closed';

  String get statusDisplay {
    switch (status) {
      case 'draft':
        return 'Draft';
      case 'active':
        return 'Active';
      case 'closed':
        return 'Closed';
      default:
        return status;
    }
  }

  bool get canVote => isActive && DateTime.now().isBefore(endDate);
  bool get hasEnded => DateTime.now().isAfter(endDate);
  
  double get turnoutPercentage => voters.isNotEmpty ? (totalVotes / voters.length) * 100 : 0;
  int get remainingVoters => voters.length - totalVotes;
  
  // Add votes property for compatibility
  List<Vote> get votes => []; // This would be populated from API
}

class Vote {
  final String voterEmail;
  final String candidateEmail;
  final DateTime votedAt;

  Vote({
    required this.voterEmail,
    required this.candidateEmail,
    required this.votedAt,
  });

  factory Vote.fromJson(Map<String, dynamic> json) {
    return Vote(
      voterEmail: json['voter_email'] ?? '',
      candidateEmail: json['candidate_email'] ?? '',
      votedAt: DateTime.parse(json['voted_at'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'voter_email': voterEmail,
      'candidate_email': candidateEmail,
      'voted_at': votedAt.toIso8601String(),
    };
  }
}

class ElectionResults {
  final int totalVotes;
  final int totalVoters;
  final double turnoutPercentage;
  final List<CandidateResult> results;
  final CandidateResult? winner;
  final bool visible;

  ElectionResults({
    required this.totalVotes,
    required this.totalVoters,
    required this.turnoutPercentage,
    required this.results,
    this.winner,
    this.visible = true,
  });

  factory ElectionResults.fromJson(Map<String, dynamic> json) {
    final visible = json['visible'] ?? true;
    final rawResults = visible
        ? (json['results'] as List<dynamic>? ?? [])
        : <dynamic>[];
    final resultsList = rawResults
        .map((r) => CandidateResult.fromJson(r))
        .toList();
    
    // Sort by votes descending
    resultsList.sort((a, b) => b.votes.compareTo(a.votes));

    final totalVoters = json['total_voters'] ?? 0;
    final totalVotes = json['total_votes'] ?? 0;
    
    return ElectionResults(
      totalVotes: totalVotes,
      totalVoters: totalVoters,
      turnoutPercentage: totalVoters > 0
          ? (totalVotes / totalVoters * 100).toDouble()
          : (json['turnout_percentage'] ?? 0.0).toDouble(),
      results: resultsList,
      winner: resultsList.isNotEmpty ? resultsList.first : null,
      visible: visible,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'total_votes': totalVotes,
      'total_voters': totalVoters,
      'turnout_percentage': turnoutPercentage,
      'results': results.map((r) => r.toJson()).toList(),
      'winner': winner?.toJson(),
    };
  }
}

class CandidateResult {
  final String name;
  final String email;
  final int votes;

  CandidateResult({
    required this.name,
    required this.email,
    required this.votes,
  });

  factory CandidateResult.fromJson(Map<String, dynamic> json) {
    return CandidateResult(
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      votes: json['votes'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'email': email,
      'votes': votes,
    };
  }

  double getPercentage(int totalVotes) {
    return totalVotes > 0 ? (votes / totalVotes) * 100 : 0;
  }
}

class Candidate {
  final String name;
  final String email;
  final String? bio;
  final String? photo;
  final int? votes;

  Candidate({
    required this.name,
    required this.email,
    this.bio,
    this.photo,
    this.votes,
  });

  factory Candidate.fromJson(Map<String, dynamic> json) {
    return Candidate(
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      bio: json['bio'],
      photo: json['photo'],
      votes: json['votes'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'email': email,
      'bio': bio,
      'photo': photo,
      'votes': votes,
    };
  }
}

class Voter {
  final String name;
  final String email;
  final bool? hasVoted;
  final DateTime? votedAt;

  Voter({
    required this.name,
    required this.email,
    this.hasVoted,
    this.votedAt,
  });

  factory Voter.fromJson(Map<String, dynamic> json) {
    return Voter(
      name: json['name'] ?? '',
      email: json['email'] ?? '',
      hasVoted: json['has_voted'],
      votedAt: json['voted_at'] != null ? DateTime.parse(json['voted_at']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'email': email,
      'has_voted': hasVoted,
      'voted_at': votedAt?.toIso8601String(),
    };
  }
}

File: backend/routes/election.js
const express = require('express');
const router = express.Router();
const Election = require('../models/Election');
const User = require('../models/User');
const Organization = require('../models/Organization');
const Notification = require('../models/Notification');
const upload = require('../middleware/upload');
const { authenticate, authorize } = require('../middleware/auth');
const { parseCSVFile, generatePassword, generateSystemEmail } = require('../utils/csvHandler');
const { sendCredentialsEmail, sendElectionNotification } = require('../utils/emailSender');
const { logActivity, ACTIONS } = require('../utils/auditLogger');
const { generateElectionResultsHTML } = require('../utils/pdfGenerator');

// Upload and parse CSV
router.post('/upload-participants', upload.single('file'), async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ message: 'No file uploaded' });
    }
    
    const result = await parseCSVFile(req.file.buffer);
    
    res.json({
      message: 'CSV parsed successfully',
      participants: result.participants,
      total: result.total,
      errors: result.errors
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Create election
router.post('/create', authenticate, authorize('organizer', 'admin'), async (req, res) => {
  console.log('=== CREATE ELECTION REQUEST ===');
  console.log('Headers:', req.headers);
  console.log('Body:', JSON.stringify(req.body, null, 2));
  console.log('User:', req.user);
  
  try {
    const {
      title,
      description,
      organizer_email,
      start_date,
      end_date,
      result_visibility,
      candidates,
      voters
    } = req.body;
    
    const user = await User.findOne({ email: organizer_email });
    if (!user) {
      return res.status(404).json({ message: 'Organizer not found' });
    }
    
    const organization = await Organization.findById(user.organization_id);
    if (!organization) {
      return res.status(404).json({ message: 'Organization not found' });
    }
    
    const election = await Election.create({
      title,
      description,
      organizer_email,
      organization_id: user.organization_id,
      start_date: new Date(start_date),
      end_date: new Date(end_date),
      result_visibility: result_visibility || 'hidden',
      candidates: [],
      voters: [],
      votes: []
    });
    
    // Create candidate accounts
    let candidateIndex = 1;
    const emailPromises = [];
    for (const candidate of candidates) {
      const systemEmail = generateSystemEmail(candidate.name, 'candidate', candidateIndex++, organization.name);
      const password = generatePassword();
      
      let candidateUser = await User.findOne({ email: systemEmail });
      if (!candidateUser) {
        candidateUser = await User.create({
          email: systemEmail,
          password,
          name: candidate.name,
          role: 'candidate',
          organization_id: user.organization_id,
          real_email: candidate.email
        });
      } else {
        // Update password so it stays in sync with the email we're about to send
        candidateUser.password = password;
        candidateUser.real_email = candidate.email;
        await candidateUser.save();
      }
      
      election.candidates.push({
        name: candidate.name,
        email: systemEmail,
        bio: candidate.bio || '',
        photo: candidate.photo || ''
      });
      
      // Always send credentials so the user has the current password
      emailPromises.push(
        sendCredentialsEmail(candidate.email, candidate.name, systemEmail, password, title, 'candidate')
          .catch(err => console.error('Email failed for candidate', candidate.email, err))
      );
    }
    
    // Create voter accounts
    let voterIndex = 1;
    for (const voter of voters) {
      const systemEmail = generateSystemEmail(voter.name, 'voter', voterIndex++, organization.name);
      const password = generatePassword();
      
      let voterUser = await User.findOne({ email: systemEmail });
      if (!voterUser) {
        voterUser = await User.create({
          email: systemEmail,
          password,
          name: voter.name,
          role: 'voter',
          organization_id: user.organization_id,
          real_email: voter.email
        });
      } else {
        // Update password so it stays in sync with the email we're about to send
        voterUser.password = password;
        voterUser.real_email = voter.email;
        await voterUser.save();
      }
      
      election.voters.push({
        name: voter.name,
        email: systemEmail
      });
      
      // Always send credentials so the user has the current password
      emailPromises.push(
        sendCredentialsEmail(voter.email, voter.name, systemEmail, password, title, 'voter')
          .catch(err => console.error('Email failed for voter', voter.email, err))
      );
    }
    
    await election.save();
    
    // Send all emails in background — don't block the response
    Promise.allSettled(emailPromises).then(results => {
      const failed = results.filter(r => r.status === 'rejected').length;
      if (failed > 0) console.error(`${failed} credential emails failed to send`);
    });
    
    // Notify organizer and log — fire-and-forget
    Notification.create({
      user_email: organizer_email,
      title: 'Election Created',
      message: `Your election "${title}" has been created with ${candidates.length} candidates and ${voters.length} voters.`,
      type: 'success',
      related_id: election._id
    }).catch(err => console.error('Notification failed:', err));

    logActivity(organizer_email, ACTIONS.ELECTION_CREATED, {
      election_id: election._id,
      election_title: title
    }, req.ip).catch(err => console.error('Audit log failed:', err));
    
    res.status(201).json({
      message: 'Election created successfully',
      election: {
        id: election._id,
        title: election.title,
        candidates: election.candidates.length,
        voters: election.voters.length
      }
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Get organizer's elections
router.get('/my-elections', authenticate, async (req, res) => {
  try {
    const { email } = req.query;
    const elections = await Election.find({ organizer_email: email });
    res.json({ elections });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Get elections by organization (admin view)
router.get('/by-organization/:orgId', authenticate, authorize('admin'), async (req, res) => {
  try {
    const elections = await Election.find({ organization_id: req.params.orgId });
    res.json({ elections });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Cast vote
router.post('/vote', authenticate, async (req, res) => {
  try {
    const { election_id, voter_email, candidate_email } = req.body;
    
    const election = await Election.findById(election_id);
    if (!election) {
      return res.status(404).json({ message: 'Election not found' });
    }
    
    if (election.status !== 'active') {
      return res.status(400).json({ message: 'Election is not active' });
    }
    
    const hasVoted = election.votes.some(v => v.voter_email === voter_email);
    if (hasVoted) {
      return res.status(400).json({ message: 'You have already voted in this election' });
    }
    
    election.votes.push({
      voter_email,
      candidate_email,
      voted_at: new Date()
    });
    election.total_votes += 1;
    await election.save();
    
    await Notification.create({
      user_email: voter_email,
      title: 'Vote Confirmed',
      message: `Your vote in "${election.title}" has been recorded.`,
      type: 'success',
      related_id: election_id
    });
    
    await logActivity(voter_email, ACTIONS.VOTE_CAST, {
      election_id,
      election_title: election.title
    }, req.ip);
    
    res.json({ success: true, message: 'Vote cast successfully' });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Export results as CSV  (must be before /results/:election_id to avoid param conflict)
router.get('/results/:election_id/export', authenticate, async (req, res) => {
  try {
    const election = await Election.findById(req.params.election_id);
    if (!election) return res.status(404).json({ message: 'Election not found' });

    const voteCounts = {};
    election.candidates.forEach(c => { voteCounts[c.email] = 0; });
    election.votes.forEach(v => {
      if (voteCounts[v.candidate_email] !== undefined) voteCounts[v.candidate_email]++;
    });

    const results = election.candidates
      .map(c => ({ name: c.name, email: c.email, votes: voteCounts[c.email] || 0 }))
      .sort((a, b) => b.votes - a.votes);

    let csv = 'rank,name,email,votes,percentage\n';
    results.forEach((r, i) => {
      const pct = election.total_votes > 0
        ? ((r.votes / election.total_votes) * 100).toFixed(1)
        : '0.0';
      csv += `${i + 1},${r.name},${r.email},${r.votes},${pct}%\n`;
    });

    res.json({
      csv,
      filename: `results_${election.title.replace(/\s+/g, '_')}.csv`,
      total_votes: election.total_votes,
      election_title: election.title
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Get election results
router.get('/results/:election_id', authenticate, async (req, res) => {
  try {
    const { election_id } = req.params;
    const { role, email } = req.query;
    
    const election = await Election.findById(election_id);
    if (!election) {
      return res.status(404).json({ message: 'Election not found' });
    }
    
    const isOrganizer = election.organizer_email === email;
    const isAdmin = role === 'admin';
    const isCandidate = election.candidates.some(c => c.email === email);
    
    let visible = false;
    
    if (election.result_visibility === 'hidden') {
      visible = isOrganizer || isAdmin || (isCandidate && election.status === 'closed');
    } else if (election.result_visibility === 'live') {
      visible = true;
    } else if (election.result_visibility === 'final_only') {
      visible = election.status === 'closed';
    }
    
    const voteCounts = {};
    election.candidates.forEach(c => {
      voteCounts[c.email] = 0;
    });
    
    election.votes.forEach(v => {
      if (voteCounts[v.candidate_email] !== undefined) {
        voteCounts[v.candidate_email]++;
      }
    });
    
    const results = election.candidates.map(c => ({
      name: c.name,
      email: c.email,
      votes: voteCounts[c.email] || 0
    })).sort((a, b) => b.votes - a.votes);
    
    res.json({
      election_id,
      title: election.title,
      total_votes: election.total_votes,
      total_voters: election.voters.length,
      visible,
      result_visibility: election.result_visibility,
      is_candidate: isCandidate,
      results: visible ? results : []
    });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Update election status
router.put('/update-status', authenticate, authorize('organizer', 'admin'), async (req, res) => {
  try {
    const { election_id, status } = req.body;
    
    const election = await Election.findById(election_id);
    if (!election) {
      return res.status(404).json({ message: 'Election not found' });
    }
    
    election.status = status;
    if (status === 'closed') {
      election.closed_at = new Date();
    }
    await election.save();
    
    if (status === 'active') {
      // Notify all voters
      for (const voter of election.voters) {
        await Notification.create({
          user_email: voter.email,
          title: 'Election Started',
          message: `The election "${election.title}" is now active. Cast your vote!`,
          type: 'info',
          related_id: election_id
        });
      }
      // Notify all candidates
      for (const candidate of election.candidates) {
        await Notification.create({
          user_email: candidate.email,
          title: 'Election Started',
          message: `The election "${election.title}" you are participating in is now active.`,
          type: 'info',
          related_id: election_id
        });
      }
      // Notify organizer
      await Notification.create({
        user_email: election.organizer_email,
        title: 'Election Activated',
        message: `Your election "${election.title}" is now live with ${election.voters.length} voters.`,
        type: 'success',
        related_id: election_id
      });
    }

    if (status === 'closed') {
      // Notify organizer
      await Notification.create({
        user_email: election.organizer_email,
        title: 'Election Closed',
        message: `Your election "${election.title}" has been closed. ${election.total_votes} votes were cast.`,
        type: 'info',
        related_id: election_id
      });
    }
    
    res.json({ message: 'Election status updated', election });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Get voter's elections
router.get('/voter-elections', authenticate, async (req, res) => {
  try {
    const { email } = req.query;
    
    const elections = await Election.find({
      'voters.email': email
    });
    
    const electionsWithVoteStatus = elections.map(election => ({
      ...election.toObject(),
      has_voted: election.votes.some(v => v.voter_email === email)
    }));
    
    res.json({ elections: electionsWithVoteStatus });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Get candidate's elections
router.get('/candidate-elections', authenticate, async (req, res) => {
  try {
    const { email } = req.query;
    
    const elections = await Election.find({
      'candidates.email': email
    });
    
    res.json({ elections });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Delete election
router.delete('/delete/:election_id', authenticate, authorize('organizer', 'admin'), async (req, res) => {
  try {
    const election = await Election.findById(req.params.election_id);
    if (!election) return res.status(404).json({ message: 'Election not found' });
    await Election.findByIdAndDelete(req.params.election_id);
    await logActivity(req.user.email, ACTIONS.ELECTION_DELETED || 'ELECTION_DELETED', {
      election_id: req.params.election_id,
      election_title: election.title
    }, req.ip);
    res.json({ message: 'Election deleted successfully' });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

// Update election details (title, description, dates, visibility)
router.put('/update/:election_id', authenticate, authorize('organizer', 'admin'), async (req, res) => {
  try {
    const { title, description, start_date, end_date, result_visibility } = req.body;
    const election = await Election.findById(req.params.election_id);
    if (!election) return res.status(404).json({ message: 'Election not found' });
    if (title) election.title = title;
    if (description) election.description = description;
    if (start_date) election.start_date = new Date(start_date);
    if (end_date) election.end_date = new Date(end_date);
    if (result_visibility) election.result_visibility = result_visibility;
    await election.save();
    res.json({ message: 'Election updated', election });
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
});

module.exports = router;
